make the jobsplitting trustsitelists aware

hufnagel opened the pull request:
If the TrustSiteLists paramater is set, change the job splitting to only look at sitewhitelists and siteblacklists and not real locations.

Currently running a Tier0 replay (PromptReco at T1) to test the basics.

@ticoann @amaltaro does the change look alright ?

hufnagel committed:
make the jobsplitting trustsitelists aware

If the TrustSiteLists paramater is set, change the job splitting to
only look at sitewhitelists and siteblacklists and not real locations.

Also do the PNN to PSN conversion in the JobCreator and pass the list
of possible directly to the JobSubmitter in the job record.

amaltaro commented:
I think you'll need to cast them to a set first.
Then we'd also need to update the workqueue code which also does data location, though I didn't check where in the code.

hufnagel commented:
We can ignore the workqueue for now, the mechanism there is completely different and orthogonal to this. All we need is to propagate the trigger for TrustSiteLists to the request level so that it can actually be used. Then we will have two different parameters that do the different ways to do xrootd overrides. We can keep them both for now.

hufnagel commented:
Actually, this approach doesn't work. Problem is that file['locations'] are PNN and sitewhitelist and siteblacklist are PSN. Of the top of my hat the only way around that is to add a PSN to PNN lookup in the job splitter to translate the sitewhitelist/siteblacklist into PNN.

This is a bit silly then though as we do a PSN to PNN translation in the JobCreator just to have a reverse PNN to PSN translation in the JobSubmitter.

I see two ways to avoid this silliness:

1) If trustsitelist is used override the job file locations with a special single "OVERRIDE' location. That will ensure that the job splitting algorithm will consider them all at the same place and do the right thing. The JobSubmitter doesn't care either because it anyways takes the possible sites to submit to from the sitewhitelist and siteblacklist, ie. it doens't even look at the job file locations.

2) Remove the PNN to PSN translation from the JobSubmitter altogether and have the JobCreator determine the PSN for a job based on the file location or the sitewhitelist/siteblacklist if trustsitelists is used.

Method 2 is a bit cleaner but has the disadvantage that the PNN to PSN matching is only done once at job creation time. This assumes though that the JobSubmitter in its current implementation would actually take into account changes over time in the PNN to PSN mapping. Does it ?

Comments ?

amaltaro commented:
I'm confused now. At the JobCreator level, what we need to have (at least for this case) is the PSN only, since we do not care about the PNN for trustSitelists. Said that, can't we put the PSN in the job 'location' for THIS specific case?

hufnagel commented:
No. The JobCreator uses input file locations to pre-sort files for job splitting. The assumption here is that jobs at different sites should never be assigned to the same job.

This isn't true anymore when you use trustsitelists. In this case the jobsplitter needs to consider these files to all be in the same location. The way the code is written, if I want to accomplish this correctly, I need to take the sitewhitelist - siteblacklist, convert that list of PSN to PNN, update the jobs input file location with these PNN and then pass them to the job splitting algorithm.

hufnagel commented:
So that brings us to the two options I mentioned.

Skip the PSN to PNN conversion I mentioned and just set the PNN for the jobs input file to some random identical value so the job splitting works correctly. We pass a non-valid PNN to the JobSubmitter as part of the job record, but it doesn't matter because the JobSubmitter skips the PNN to PSN conversion if trustsitelists is used.

Or do the PSN to PNN conversion in the JobSubmitter, but if we have to have the code to do that there anyways, I would rather do everything there and just determine where we can submit jobs to in the JobCreator and pass that list of possible PSN to the JobSubmitter (and remove the PNN to PSN conversion from the JobSubmitter).

amaltaro commented:
> The assumption here is that jobs at different sites should never be assigned to the same job.

You meant files in different sites should not be assigned to the same job, right? However, this is AAA world, we don't know and we *should* not care about data/file location.

hufnagel commented:
Well, the job splitter cares !

ticoann commented:
I think  Dirk's method 2 is cleaner. 
By converting file location (supposed to be PNN) to PSN here. (Not completely sure that is the only place)
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/JobSplitting/JobFactory.py#L212
and remove the conversion from 
https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py#L323

So split job only handles PSN. I don't think we should concern about PSN and PNN mapping change between job creation and job submission.  That mapping doesn't change often write?

However current patch should work as well. It will simply create the splitting correctly when the trusted sites are set but that location is not used for submission anyway since we that location is completely ignored.

Personally I prefer cleaner changes (Dirk's 2nd method) if the change is simple. But we have to be careful whether there will be any hidden side effect.

ticoann commented:
Of course this assumes we have the site populated in wmbs_location table. (Current code won't populate the sites without pnn in site db).

hufnagel commented:
I am testing my method 1 right now because it was faster to create the code. But I also prefer method 2 and will look into it.

ticoann commented:
Dirk, do you really need to do fake site "aaa"? Doesn't your initial patch work? Either you set the fake location or not it should be ignored, isn't it?

hufnagel commented:
I need to because the available files are sorted according to locSet. They need to have the same locSet for the splitting algorithm to all process them together.

ticoann commented:
I mean instead of using 'aaa', use actual white list site although it is not PNN but PSN.

hufnagel commented:
Functionally it should be the same, if its not and these file locations are actually accessed in the trustsitelists case I want to find this out with this test (as they shouldn't be).

hufnagel commented:
Ok, I am implementing method 1 now. One thing I noticed, the system assumes that site whitelist/blacklist are in WMAgent siteName, which is internally different than cmsName. I think in practice everyone assumes that site whitelist/blacklist is in cmsName, ie. PSN anyways, so I removed that conversion code from the JobSubmitter.

Way I implemented this to keep the JobCreator code simple is to take the file locations, convert them to PSN, then sort by them but also update the file locations to the PSN so they get included with the job record that is written out later and passed to the JobSubmitter. Then the JobSubmitter just interprets these file locations as possible PSN it can run at and just modified according to sites that are in draining or down.

First try, almost expect some unit tests to break, but want to see what jenkins makes of it.

hufnagel commented:
Ok, that doesn't work. The Tier0 doesn't use sortByLocation() because it has to do its own data discovery queries, therefore anything implemented there doesn't work. I need to look at what the JobCreator does with the jobgroups and jobs it gets back from the splitting algorithm.

Which means for sortByLocation() I am going back to just using "AAA" as a possible location.

hufnagel commented:
retest this please

hufnagel commented:
Ok, that last commit worked (the testProductionMultiQueue failure is bogus). I removed the site whitelis and blacklist functionality test from the JobSubmitter unit test though, need to add at least one test for this to one of the job splitting tests to make sure it works correctly. Will use the FileBased splitter unit test (doing it in one is enough).

hufnagel commented:
I also still want to remove the job input files location data from the job object that is passed to the JobSubmitter. It currently breaks some of the unit tests because the File objects are reused multiple times (first test works then, later break because the location is gone). Will need to do some refactoring to avoid that reuse.

hufnagel commented:
retest this please

ericvaandering commented:
test this please

hufnagel commented:
Ok, great, unit tests are happy :-). @ticoann @amaltaro , can you review please ?

I am currently running another Tier0 replay, but its almost done and I didn't see any problems.

hufnagel commented:
One question I still have, what is the difference between siteName and cmsName ? Is there one in practice. It seems to me we use siteName as PSN (because that is what the JobSubmitter is using as location. The site whitelist/blacklist were assumed to be cmsNames though.

Are they different anywhere in our system ? I though what the requestors put into site whitelist/blacklist is actually a PSN, ie. a location jobs are submitted to...

hufnagel commented:
This patch changes the old behavior in that it treats site whitelist/blacklist as siteName now, ie. they are interpreted directly as "sites" that are used by the submission system and do not go through a cmsName ->siteName conversion first. Would be good if someone can confirm that this won't cause problems.

amaltaro commented:
I see we have both cms_name and site_name in the wmbs_location table, though all the 125 rows/sites in my agent have exactly the same value for them. I don't think there will be an issue here.

Now the question could be whether their definition is the same or not, I honestly have no idea.

hufnagel commented:
I think the design has a separation of what a "site" is called within WMAgent and what a CMS "site" name is, that's why we have a conversion from the cmsName in site whitelist/blacklist to site name. In practice though that same site name is then used directly for job submission.

So while you could setup a site with siteName "site1" with cmsName "T1_US_FNAL", the old code would correctly handle the conversion of "T1_US_FNAL" to "site1", in practice this is useless because this concept of "site1" does not extend to gWMS. If you look at

WMCore/src/python/WMCore/WMBS/MySQL/Locations/GetPSNtoPNNMapping.py

the siteName is directly interpreted as a PSN and we later put it into the JDL to request what sites a job can run at.

So I think in practice cmsName is useless. Agreed ?

amaltaro commented on src/python/WMCore/JobSplitting/JobFactory.py:
```diff
@@ -27,11 +24,11 @@ class JobFactory(WMObject):
         self.package = package
         self.subscription  = subscription
-        self.generators    = generators
+        self.generators    = generators if generators else []
```
Besides shutting up pylint, it makes no sense. Why don't we just default to [] since lambda function is called on it later on?

hufnagel replied:
Well, the usual argument about not having empty lists or dicts as default arguments applies. We discussed this before, its just in general a really bad habit and you should never do it, even if in the specific case it doesn't hurt you.

amaltaro replied:
That's why I made my comment, because implicitly the default value is still []  
If we default to [] 3 lines below the previous location, then just leave it as it was before :)

hufnagel replied:
They are different. The new code creates a new empty list every time the function is called. With the [] in the call definition that is not the case.

amaltaro commented:
> So I think in practice the separation between siteName and cmsName is useless. Agreed ?

I think so, though I still see cms_name being used in a few other places here. Perhaps we should also update them to site_name to avoid confusion?

hufnagel commented:
So I think in practice the separation between siteName and cmsName is useless. Agreed ?

    I think so, though I still see cms_name being used in a few other places here. Perhaps we should also update them to site_name to avoid confusion?

Yes, agreed. But I would open a new issue/PR for that, this PR is already long enough.

hufnagel commented:
I just noticed a comment I need to change, but I want to wait until the rest of the code is reviewed and has been signed off on. I can then fix the comment, update the PR and we can merge after the jenkins run completes again.

amaltaro commented:
Hmm, it seems the underlying plugins are using either "locations" or "possibleSites" key to get the PSN (target site list) for submission, see
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/BossAir/Plugins/CondorPlugin.py#L1010

how did your test work if it was removed from JobSumitterPoller code (?)

hufnagel commented:
If you look at the JobSubmitter code, it generates possibleSites from possiblePSN.

https://github.com/hufnagel/WMCore/blob/a4e274b1fe677622b1c5439aae271de915d0327e/src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py#L292

https://github.com/hufnagel/WMCore/blob/a4e274b1fe677622b1c5439aae271de915d0327e/src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py#L369

https://github.com/hufnagel/WMCore/blob/a4e274b1fe677622b1c5439aae271de915d0327e/src/python/WMComponent/JobSubmitter/JobSubmitterPoller.py#L665

Seems to me that possibleSites is directly based on possiblePSN (maybe we can unify somehow ?) and potentialLocations is based on that plus the filter on down/drained sites.

hufnagel commented:
Look here

https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/BossAir/Plugins/CondorPlugin.py#L1018

DESIRED_Sites gets all possible sites ignoring site down/drain and ExtDESIRED_Sites gets the list of sites that includes site down/drain ? Does this make sense ?

ticoann commented:
> DESIRED_Sites gets all possible sites ignoring site down/drain and ExtDESIRED_Sites gets the list of sites that includes site down/drain ? Does this make sense ?

I might be wrong but I don't think ExtDESIRED_Sites is used in Glidein at least for now. Do you know how ExtDESIRED_Sites is used?

hufnagel commented:
Well, if we don't use EstDESIRED_Sites, we don't use site down/drain status during job submission and we might as well rip that code out... It might have been used ages ago when we were using other submission plugins, but for the CondorPlugin the sit down/drain status seems to only affect ExtDESIRED_Sites.

hufnagel commented:
But again, we should open a new issue for that :-). My changes don't touch this part.

ticoann commented:
Hi Dirk,
One thing I am concerning is that now the trusted/white site list matching/assign is happening in job creation instead of job submission. One outstanding issue we had is that add ability to make site whitelist change after workflow is running status. I think it wouldn't make big difference if the time between job creation and submission is small but it would be better that matching happens in submission time than creation time.

ticoann commented:
I think we eventually clean up all other location handling other (i.e. CMS site)  than PSN and PNN.  But we could handle that in separate issue.

hufnagel commented:
> One thing I am concerning is that now the trusted/white site list matching/assign is happening in job creation instead of job submission. One outstanding issue we had is that add ability to make site whitelist change after workflow is running status. I think it wouldn't make big difference if the time between job creation and submission is small but it would be better that matching happens in submission time than creation time.

Is that really a problem ? You anyways will have to deal with jobs that are already running/pending, there is no way to change them unless you condor edit them. So changing the whitelist or blacklist will never be instantaneous. 

I could change this back, but it will make the code less clean because you have to propagate this information further down and can't deal with all of these dependencies in one place. My changes you evaluate jobs input file locations, the associated PSN, the site whitelist/blacklist and whether we should ignore file location and trust the site whitelist/blacklist in a single piece of code. If you apply the site whitelist/blacklist during job submissions, you do it in two places.

Also would have to revert/change a few unit tests that rely on this being done in the JobCreator now.

Its doable, but is it worth it, given that even with this change you still won't be able to switch site whitelist/blacklist immediately, there will always be a backlog of jobs that you can only edit in condor ?

hufnagel commented:
Btw, if you want to change the site whitelist/blacklist for already created jobs, you actually would have to edit the job sandbox that is written to disk. This information isn't passed via database, its persistent in files on disk.

Are we really doing this ? If we don't, we might as well do everything in the JobCreator because that is anyway as far as an updated site whitelist/blacklist will reach.

hufnagel commented:
Just had a chat with @ericvaandering about this, I don't think we want to support automatic patching of job sandboxes. And for manual Ops procedures/scripts you can still change the possiblePSN parameter in the job sandbox.

ticoann commented:
Hi Dirk,  All your concern is valid. that is why we didn't implemented yet. 

The main reason this was issue was when workflow is created site was fine but ops want to be able to change the site whitelist when they found the problem in the middle of running. (better site comes out in terms of resource) 

> Is that really a problem ? You anyways will have to deal with jobs that are already running/pending, there is no way to change them unless you condor edit them. So changing the whitelist or blacklist will never be instantaneous.

running jobs are out of control until it fails and comes back. For pending jobs, we can use condor_edit to change but I really don't want to do that.  

> Btw, if you want to change the site whitelist/blacklist for already created jobs, you actually would have to edit the job sandbox that is written to disk. Every single job sandbox.

Yes, we have to change that as well.

So initial plan was we don't  care about already submitted jobs. But we change this before submission.
Anyway, since Eric said not supporting this. :smile:  I am fine with that.

hufnagel commented:
Its not that we will never support this. But in the current design it doesn't make sense to have an automatic mechanism that modifies every single job sandbox on disk. We might still have to support that functionality, but not in an automated fashion, but only for emergency "these jobs won't run anywhere if we don't patch the sandbox" manual interventions. And for these you can always modify possiblePSN and directly specify where jobs should be run, no need to fiddle with whitelist or blacklist.

When we do the planned redesign with the division of "unit of work" and job, we should keep this use case in mind though, maybe there is an easier way to accomplish this...

hufnagel commented:
Ok, as there weren't any further code comments I am fixing the one comment and push the changes. Once this goes through another jenkins test we can decide when this should be merged.

ticoann commented:
test this please

ericvaandering commented:
test this please
